feat: Migrate Table to fiori3

JKMarkowski opened the pull request:
## Related Issue
Related to https://github.com/SAP/fundamental-styles/issues/914

## Descriptions
There are included
- Base Table Component - Header, Body, Footer + sizes
- Interactive states(hover, active, selected)
- Table with checkboxes + sizes
- Highlight Row
- Table with pagination
- Responsive table + pop-in mode
- Filter / Sort of table
- Navigation indicator inside table
- Fixed Column

## Breaking Changes
Added Class:
Modifiers:
- `fd-fable` + `(__header/__body/__cell)` + `(--no-vertical-borders/--no-horizontal-borders)`
- `fd-table` + `(__row/_cell)` + `(--activable/--hoverable)`
- `fd-table__cell--checkbox`

- `fd-table__cell` + `(--success/--error/--warning--information)`
- `fd-table__cell--status-indicator`+ `(--success/--error/--warning--information)`


Responsive:
- `fd-table__row--main`
- `fd-table__row--secondary`


Removed Classes:
Elements:
- `fd-table__sort-column(--asc/--desc)`
- `fd-table__context-menu`
- `fd-table__context-menu-label`
Modifiers: 
- `fd-table__row(--success/--error/--warning/--information)`



## Screenshots
> **NOTE:** If you've made any style changes, please provide appropriate screenshots (before and after) to help reviewers.
>
> To see examples of which screenshots to include, go to [Screenshot Examples](https://github.com/SAP/fundamental-styles/wiki/Pull-Request-Screenshot-Examples).

### Before:
![image](https://user-images.githubusercontent.com/26483208/82561645-032cca00-9b74-11ea-95e7-8d64c313c65b.png)
![image](https://user-images.githubusercontent.com/26483208/82561656-07f17e00-9b74-11ea-9c0e-b16f68e7c977.png)


### After:
![image](https://user-images.githubusercontent.com/26483208/82561703-1c357b00-9b74-11ea-843a-1a7e480c7edd.png)

![image](https://user-images.githubusercontent.com/26483208/82561680-0fb12280-9b74-11ea-814d-c88aa08f8442.png)

#### Please check whether the PR fulfills the following requirements

- [x] all items on the PR Review Checklist are addressed :
https://github.com/SAP/fundamental-styles/wiki/PR-Review-Checklist

mikerodonnell89 commented:
Too many cells in the footer on this example
![Screen Shot 2020-05-21 at 1 24 21 PM](https://user-images.githubusercontent.com/2471874/82597674-95c66c80-9b66-11ea-8bbe-2e8788809e11.png)

mikerodonnell89 commented:
Strange focus outline for condensed checkboxes
![Screen Shot 2020-05-21 at 1 24 21 PM](https://user-images.githubusercontent.com/2471874/82598284-a62b1700-9b67-11ea-94da-94beae444162.png)

mikerodonnell89 commented:
Table cell contents aligned left in RTL
![Screen Shot 2020-05-21 at 1 38 31 PM](https://user-images.githubusercontent.com/2471874/82598802-6fa1cc00-9b68-11ea-91f6-8c3d34b018bd.png)

mikerodonnell89 commented on src/table.scss:
```diff
@@ -11,32 +11,29 @@
-  --fd-table-header-background-color: var(--fd-color-neutral-1);
+
+  $fd-table-cell-padding: 0.5rem;
+  $fd-table-header-height: 2.75rem;
```
Weird name for this variable because it is only used in cell selectors

JKMarkowski replied:
Naming changed

mikerodonnell89 commented on src/table.scss:
```diff
@@ -48,258 +45,173 @@ $block: #{$fd-namespace}-table;
-    @include fd-hover() {
-      @include fd-var-color("background-color", $fd-table-row-background-color--hover, --fd-color-background-hover);
+      &:hover {
+        background-color: inherit;
```
is this needed for something that I'm missing?

InnaAtanasova commented on src/table.scss:
```diff
@@ -48,258 +45,173 @@ $block: #{$fd-namespace}-table;
+    padding: 0 $fd-table-cell-padding;
+    font-family: var(--sapFontFamily);
+    font-size: var(--sapFontSize);
+    font-weight: normal;
```
normal font-weight is coming with the fd-reset

InnaAtanasova commented on src/table.scss:
```diff
@@ -48,258 +45,173 @@ $block: #{$fd-namespace}-table;
 
-  &__row {
-    transition: background-color $fd-table-transition-params;
+  &__header {
```
I think your header is missing the border

InnaAtanasova commented on src/table.scss:
```diff
@@ -48,258 +45,173 @@ $block: #{$fd-namespace}-table;
+    font-size: var(--sapFontSize);
+    font-weight: normal;
+    color: inherit;
+    vertical-align: middle;
```
I don't think this is doing anything

JKMarkowski commented:
Hi @InnaAtanasova 
> 1. I don't think the cells should have active state. They have hover, selected, hover+selected but didn't see the active (I might have missed it). The cells get focus tho (seen in the demo app)
- There is down state, which I understand as a active. You can take a look at activable cells/rows example on preview.

> 2. The RTL support is "missing" - the text should not be left-aligned and the first column padding (1rem) should go to the other side
- Fixed

> 3. Items that are navigatable have different color
>    ![Screen Shot 2020-05-21 at 5 18 46 PM](https://user-images.githubusercontent.com/39598672/82608084-fe721100-9b87-11ea-9cfb-ada1fa103146.png)
- It's shown only on some images, I don't know what to do with links actually links got own colors. Maybe I should have something like `fd-table__text` to keep color inherited.

> 1. Freeze - shouldn't we have a horizontal/vertical scroll when we freeze a certain row or column? For example, if we freeze the first column, we will be able to horizontally scroll in order to see all the columns (this is usually when we can't fit all the columns in the vw)

We have Fixed Column now. It's in last example, I would also add something like fixed row.

salarenko commented:
Nice work.
I'm not sure if this composition approach is correct in terms of interactivity and other features. Solid base should be a table with basic styles. Currently we bring with base table additional behaviors like:
- Marking row on hover
- Interactive table headers
I'd rather consider those as extensions of basic table.

prsdthkr commented:
Great work :D
I noticed that the table here behaves like a [responsive table](https://ui5.sap.com/#/entity/sap.m.Table). If we were to follow the spec for the [grid table](https://ui5.sap.com/#/entity/sap.ui.table.Table), the columns will not be responsive by default as seen in [this example](https://ui5.sap.com/#/entity/sap.ui.table.Table/sample/sap.ui.table.sample.Basic). Instead, a scrollable view is provided (in smaller viewports or at 200% zoom). Also there are 3 modes for column width behavior as [seen here](https://ui5.sap.com/#/entity/sap.ui.table.Table/sample/sap.ui.table.sample.Resizing): static, flexible, and mixed.

I understand if this is out of scope for this PR.

JKMarkowski commented:
Hi @prsdthkr,

Those features with responsive behaviour are up to application specification. If user adds some `min-width` to cells and adds wrapper with `overflow-x` it's possible to achieve. I added example with fixed column, which actually shows how it can be used.

Also there is responsive table example with `pop-in` mode, which changes view of table(it has modified markup a little bit)
![image](https://user-images.githubusercontent.com/26483208/83135703-873c0000-a0e6-11ea-9f3e-d4918c0c5151.png)

I'm not sure what about the static, flexible, and mixed. It's not mentioned anywhere in fiori3 specs. We can add this example with dummy cells in future. Thanks for taking a look at it.

InnaAtanasova commented:
Documentation:
The checkbox needs to be checked with selected state
![Screen Shot 2020-05-28 at 3 12 46 PM](https://user-images.githubusercontent.com/39598672/83184138-ffaabd00-a0f6-11ea-8afd-d3218543c552.png)

Responsive table:
- missing border 
![Screen Shot 2020-05-28 at 3 11 56 PM](https://user-images.githubusercontent.com/39598672/83184221-20731280-a0f7-11ea-892d-2f940c83ad8d.png)

- prob not achievable with pure css but the whole item should get hover and selected state:
![Screen Shot 2020-05-28 at 3 11 08 PM](https://user-images.githubusercontent.com/39598672/83184291-400a3b00-a0f7-11ea-9d61-da9c0cdff8e1.png)
![Screen Shot 2020-05-28 at 3 10 46 PM](https://user-images.githubusercontent.com/39598672/83184292-400a3b00-a0f7-11ea-88e6-924741d70acf.png)

Navigation:
- I think the button should not have states. The states should be applied to the cell
![Screen Shot 2020-05-28 at 3 08 19 PM](https://user-images.githubusercontent.com/39598672/83184316-48fb0c80-a0f7-11ea-9c04-75adfc3ec71a.png)

InnaAtanasova commented on src/table.scss:
```diff
@@ -3,355 +3,415 @@
-  border-width: $fd-table-border-width;
+    &--navigation {
+      font-size: var(--sapFontSmallSize);
+      width: 2rem;
+      min-width: 2rem;
```
should be put in a variable

InnaAtanasova commented on src/table.scss:
```diff
@@ -3,355 +3,415 @@
-        }
-      }
+    &--checkbox {
+      width: 10px;
```
rem value

InnaAtanasova commented on src/table.scss:
```diff
@@ -3,355 +3,415 @@
-    &--warning {
-      @include fd-var-color("background-color", fd-color('background', 4), --fd-color-background-4);
+    &--fit-content {
+      width: 10px;
```
rem value

InnaAtanasova commented on src/table.scss:
```diff
@@ -3,355 +3,415 @@
-        }
-      }
+    &--no-padding {
+      padding: 0;
+      margin: 0;
```
since we see this twice in the component you can create a local reset mixin and reuse it

InnaAtanasova commented on src/table.scss:
```diff
@@ -3,355 +3,415 @@
+      left: 0;
+      position: absolute;
+      background-color: $fd-table-basic-background;
+      display: flex;
+      align-items: center;
```
`fd-flex-vertical-center` mixin

InnaAtanasova commented:
`--borderless-x` and `--borderless-y` instead of `--no-vertical-borders` and `--no-horizontal-borders` ? But we better discuss it with the rest of the team

InnaAtanasova commented:
![Screen Shot 2020-05-28 at 3 43 16 PM](https://user-images.githubusercontent.com/39598672/83186108-0129b480-a0fa-11ea-8912-853d791ac73a.png)

The navigated indicator should be on the other side and the icon should point to the other direction

InnaAtanasova commented:
![Screen Shot 2020-05-28 at 3 42 48 PM](https://user-images.githubusercontent.com/39598672/83186526-a47ac980-a0fa-11ea-9a05-5352fa3fa436.png)

InnaAtanasova commented:
![Screen Shot 2020-05-28 at 3 41 51 PM](https://user-images.githubusercontent.com/39598672/83186550-ae043180-a0fa-11ea-8444-8e6c03c6e5ba.png)

In RTL the fixed column should go on the right side

InnaAtanasova commented:
![Screen Shot 2020-05-28 at 3 40 33 PM](https://user-images.githubusercontent.com/39598672/83186602-c4aa8880-a0fa-11ea-93a3-d85abb54c391.png)

RTL doc example - arrows should be mirrored

JKMarkowski commented:
@InnaAtanasova 

> Documentation:
> The checkbox needs to be checked with selected state
> ![Screen Shot 2020-05-28 at 3 12 46 PM](https://user-images.githubusercontent.com/39598672/83184138-ffaabd00-a0f6-11ea-8afd-d3218543c552.png)

Done

> Responsive table:
> 
> * missing border
>   ![Screen Shot 2020-05-28 at 3 11 56 PM](https://user-images.githubusercontent.com/39598672/83184221-20731280-a0f7-11ea-892d-2f940c83ad8d.png)
It shouldn't have border (
![image](https://user-images.githubusercontent.com/26483208/83400493-36d4e300-a403-11ea-8cfa-cd41b0aafc51.png)
![image](https://user-images.githubusercontent.com/26483208/83400517-418f7800-a403-11ea-9ecf-ca4e7326f0ca.png)


> * prob not achievable with pure css but the whole item should get hover and selected state:
>   ![Screen Shot 2020-05-28 at 3 11 08 PM](https://user-images.githubusercontent.com/39598672/83184291-400a3b00-a0f7-11ea-9d61-da9c0cdff8e1.png)
>   ![Screen Shot 2020-05-28 at 3 10 46 PM](https://user-images.githubusercontent.com/39598672/83184292-400a3b00-a0f7-11ea-88e6-924741d70acf.png)
TODO

> 
> Navigation:
> 
> * I think the button should not have states. The states should be applied to the cell
>   ![Screen Shot 2020-05-28 at 3 08 19 PM](https://user-images.githubusercontent.com/39598672/83184316-48fb0c80-a0f7-11ea-9c04-75adfc3ec71a.png)
I don't think so, User shouldn't perform action on triggering cell, but button

stefanoScalzo commented:
![Screen Shot 2020-06-01 at 10 10 31 AM](https://user-images.githubusercontent.com/50607147/83417488-29fac400-a3f0-11ea-89eb-8856231a4fd5.png)
The toolbar border is overlapping with the table border

JKMarkowski commented:
@stefanoScalzo Should I be considered as something wrong?

stefanoScalzo commented:
> @stefanoScalzo Should I be considered as issue?

True, something to discuss because I think we need to give our components the option to not include the border for cases like this? either that or we use a class that positions the elements lower with a smaller z index to hide it, but this may not work due to some using box-shadow. We need to discuss it.

stefanoScalzo commented:
For: Responsive Table - Pop-in Mode with Checkboxes
When clicking on the checkbox the whole table gets highlighted but when clicking the table nothing gets checked.

JKMarkowski commented:
> For: Responsive Table - Pop-in Mode with Checkboxes
> When clicking on the checkbox the whole table gets highlighted but when clicking the table nothing gets checked.

there is no way to make it work on pure css, I can't wrap row/cell to label.

stefanoScalzo commented:
Can you include Navigated Item Indicator Bar in your responsive table example please

stefanoScalzo commented:
![Screen Shot 2020-06-01 at 11 09 51 AM](https://user-images.githubusercontent.com/50607147/83423043-6d593080-a3f8-11ea-8add-67e0ed02abad.png)
Overflow property not working properly

mikerodonnell89 commented:
Filter section has compact buttons but the rest of the table is cozy - intentional?

JKMarkowski commented:
@mikerodonnell89 It's done by accident, I will change it

InnaAtanasova commented:
> 
> > Responsive table:
> > 
> > * missing border
> >   ![Screen Shot 2020-05-28 at 3 11 56 PM](https://user-images.githubusercontent.com/39598672/83184221-20731280-a0f7-11ea-892d-2f940c83ad8d.png)
> 
> It shouldn't have border
> 
> ![image](https://user-images.githubusercontent.com/26483208/83400493-36d4e300-a403-11ea-8cfa-cd41b0aafc51.png)
> ![image](https://user-images.githubusercontent.com/26483208/83400517-418f7800-a403-11ea-9ecf-ca4e7326f0ca.png)
> 

Sorry, I meant the border on the right, half of it has a border, the other half doesn't

InnaAtanasova commented:
Impressive job with the table! Once Travis is happy you can go ahead and merge it

salarenko commented:
Looks good. I'm not sure about one visual thing. Shouldn't the checkbox in collapsed table be aligned to the top? The docs say nothing about it, but thats the case @InnaAtanasova had to cover in very similar List with checkbox layout and it looks much better.

This is how it looks now:
![image](https://user-images.githubusercontent.com/17496353/83617877-ffdd0980-a589-11ea-9e66-ca360598f910.png)

This is how it looks with checkbox aligned to the top: (`vertical-align`)
![image](https://user-images.githubusercontent.com/17496353/83617688-bbea0480-a589-11ea-8095-a90d3fb59b2c.png)
